tools/rustfmt/srcs/main.rs

```diff
@@ -0,0 +1,207 @@
+        .spawn()
+        .expect("Failed to spawn command");
+
+    let output = child
+        .wait_with_output()
+        .expect("Failed to wait on spawned command");
+
+    if !output.status.success() {
+        std::process::exit(output.status.code().unwrap_or(1));
+    }
+}
+
+/// Run rustfmt on a set of Bazel targets
+fn apply_rustfmt(options: &Config, targets: &Vec<String>) {
+    // Ensure the targets are first built and a manifest containing `rustfmt`
+    // arguments are generated before formatting source files.
+    build_rustfmt_targets(&options, &targets);
+
+    for target in targets.iter() {
+        // Replace any `:` characters and strip leading slashes
+        let target_path = target.replace(":", "/").trim_start_matches("/").to_owned();
+
+        // Load the manifest containing rustfmt arguments
+        let rustfmt_config = parse_rustfmt_manifest(
+            &options
+                .workspace
+                .join("bazel-bin")
+                .join(format!("{}.rustfmt", &target_path)),
+        );
+
+        // Ignore any targets which do not have source files. This can
+        // occur in cases where all source files are generated.
+        if rustfmt_config.sources.is_empty() {
+            continue;
+        }
+
+        // Run rustfmt
+        let status = Command::new(&options.rustfmt)
+            .current_dir(&options.workspace)
+            .arg("--edition")
+            .arg(rustfmt_config.edition)
+            .args(rustfmt_config.sources)
+            .status()
+            .expect("Failed to run rustfmt");
+
+        if !status.success() {
+            std::process::exit(status.code().unwrap_or(1));
+        }
+    }
+}
+
+/// A struct containing details used for executing rustfmt.
+#[derive(Debug)]
+struct Config {
+    /// The path of the Bazel workspace root.
+    pub workspace: PathBuf,
+
+    /// The Bazel executable to use for builds and queries.
+    pub bazel: PathBuf,
+
+    /// The rustfmt binary from the currently active toolchain
+    pub rustfmt: PathBuf,
+
+    /// The rustfmt config file containing rustfmt settings.
+    /// https://rust-lang.github.io/rustfmt/
+    pub config: PathBuf,
+
+    /// An optional command line flag used to control what targets
+    /// to format. Users are expected to either pass a Bazel label
+    /// or a package pattern (`//my/package/...`).
+    pub package: Option<String>,
+}
+
+/// Parse command line arguments and environment variables to
+/// produce config data for running rustfmt.
+fn parse_args() -> Config {
+    Config{
+        workspace: PathBuf::from(
+            env::var("BUILD_WORKSPACE_DIRECTORY")
+            .expect("The environment variable BUILD_WORKSPACE_DIRECTORY is required for finding the workspace root")
+        ),
+        bazel: PathBuf::from(
+            env::var("BAZEL_REAL")
+            .unwrap_or_else(|_| "bazel".to_owned())
+        ),
+        rustfmt: PathBuf::from(env!("RUSTFMT"))
+            .canonicalize()
```

Comment:
Reading through symlinks can cause problems in some environments which are symlink-heavy - is there a reason to canonicalize rather than using something like:
```rust
fn absolutify_existing(path: PathBuf) -> std::io::Result<PathBuf> {
    let path = if path.is_absolute() {
        path
    } else {
        std::env::current_dir().expect("Failed to get working directory").join(path)
    };
    std::fs::metadata(&path).map(|_| path)
}
```

Reply:
Thanks! I'll burn this to memory. I remember you giving similar feedback on an earlier PR and thought this was a good idea then 😅 

Updated